File: app/src/main/kotlin/com/usagelimits/core/sync/SyncEngine.kt
package com.usagelimits.core.sync

import com.usagelimits.core.auth.CredentialStore
import com.usagelimits.core.auth.OAuthCredentials
import com.usagelimits.core.database.UsageRepository
import com.usagelimits.core.model.ProviderAccount
import com.usagelimits.core.notifications.NotificationEvaluator
import com.usagelimits.core.model.SnapshotStatus
import com.usagelimits.core.model.UsageSnapshot
import com.usagelimits.core.network.ProviderException
import com.usagelimits.providers.ProviderRegistry
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.withContext
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/** Per-account outcome of a sync pass. */
data class SyncOutcome(
    val accountId: String,
    val success: Boolean,
    val message: String? = null,
)

/**
 * Refreshes usage for accounts.
 *
 * Three properties matter here:
 *
 *  - **Isolation.** Accounts sync concurrently and each failure is caught per account, so one
 *    revoked token or one provider outage cannot stop the others from updating.
 *  - **Refresh safety.** Token refreshes are serialised per credential reference. Without
 *    that, a periodic sync racing a manual pull-to-refresh could both refresh the same
 *    account, and providers invalidate the old refresh token on use — the loser would be left
 *    holding a dead token and the account would appear revoked.
 *  - **Stale over empty.** A failed fetch keeps the previous numbers and records the error, so
 *    the UI can say "last updated 37 minutes ago, refresh failed" instead of blanking a card.
 */
class SyncEngine(
    private val repository: UsageRepository,
    private val credentialStore: CredentialStore,
    private val registry: ProviderRegistry,
    private val nowMs: () -> Long = System::currentTimeMillis,
) {

    private val refreshMutexes = mutableMapOf<String, Mutex>()
    private val mutexGuard = Mutex()

    /**
     * One fetch-through-save at a time per account.
     *
     * Two passes over one account — the periodic worker and a pull-to-refresh, say — used to
     * race from fetch to save with nothing between them. A fetch that returned 80 % and was
     * slow to complete could land AFTER a later fetch that returned 5 %, and because
     * `fetchedAt` is stamped at completion the older figure arrived wearing the newer time:
     * the account rose from 5 % to 80 % with a reassuring fresh timestamp. Wall-clock
     * completion time is not an ordering. Holding this lock from fetch to save makes the
     * order of starts the order of writes, so the later attempt's figures are the ones kept.
     */
    private val accountLocks = mutableMapOf<String, Mutex>()

    private suspend fun accountLock(accountId: String): Mutex =
        mutexGuard.withLock { accountLocks.getOrPut(accountId) { Mutex() } }

    /** Syncs every account concurrently. Never throws; failures surface per account. */
    suspend fun syncAll(): List<SyncOutcome> = coroutineScope {
        repository.accounts()
            .map { account -> async { syncAccount(account) } }
            .awaitAll()
    }

    suspend fun syncAccount(accountId: String): SyncOutcome {
        val account = repository.account(accountId)
            ?: return SyncOutcome(accountId, success = false, message = "Account not found")
        return syncAccount(account)
    }

    suspend fun syncAccount(account: ProviderAccount): SyncOutcome = accountLock(account.localId).withLock {
        // Emulator fixtures never contact a provider. R8 removes this branch from release APKs.
        if (com.usagelimits.BuildConfig.DEBUG && account.attributes["screenshotFixture"] == "true") {
            repository.snapshot(account.localId)?.let { repository.saveSnapshot(it.copy(fetchedAt = nowMs())) }
            return@withLock SyncOutcome(account.localId, success = true)
        }
        try {
            val provider = registry.forId(account.provider)
                ?: throw ProviderException.Unexpected("No provider for ${account.provider.id}")

            val credentials = validCredentials(account)
            val result = try {
                provider.fetchUsage(account, credentials)
            } catch (e: ProviderException.Unauthorized) {
                // The reactive half of the refresh policy, and what makes the proactive half
                // safe to keep conservative. The provider has said this token is dead, which
                // beats any expiry the account did or did not carry.
                //
                // Once only: if the renewed pair is refused too, the credential is genuinely
                // revoked and a second attempt is just another way to fail.
                val renewed = validCredentials(account, rejectedAccessToken = credentials.accessToken)
                provider.fetchUsage(account, renewed)
            }

            repository.saveSnapshot(
                UsageSnapshot(
                    accountId = account.localId,
                    fetchedAt = nowMs(),
                    status = SnapshotStatus.OK,
                    windows = result.windows,
                    resetCredits = result.resetCredits,
                    resetCreditCount = result.resetCreditCount,
                    applicableResetCreditCount = result.applicableResetCreditCount,
                ),
            )
            SyncOutcome(account.localId, success = true)
        } catch (e: CancellationException) {
            // Cancellation is not a sync failure and must never be recorded as one.
            //
            // `CancellationException` IS an `Exception` on the JVM, so the catch below used to
            // swallow it: WorkManager stopping a periodic pass — a constraint lost, doze, the
            // system reclaiming the worker — wrote "Unexpected error while refreshing" onto
            // every account still in flight. Accounts that were perfectly healthy came back as
            // error cards, and swallowing it also breaks structured concurrency, since the
            // coroutine carries on after its scope has been cancelled.
            //
            // `recordFailure` already rethrew it for exactly this reason; this is the same
            // hazard one level up.
            throw e
        } catch (e: ProviderException) {
            // A cancelled call reaches here too, wearing the wrong clothes: OkHttp answers
            // cancellation with IOException("Canceled"), which the client maps to Offline. So
            // the scope is checked before anything is persisted — otherwise a cancelled sync
            // records "No network connection" against an account with a working connection.
            currentCoroutineContext().ensureActive()
            recordFailure(account.localId, e.userMessage(), e is ProviderException.Unauthorized)
            SyncOutcome(account.localId, success = false, message = e.userMessage())
        } catch (e: Exception) {
            // A provider bug must not take the whole sync pass down with it.
            currentCoroutineContext().ensureActive()
            val message = "Unexpected error while refreshing"
            recordFailure(account.localId, message)
            SyncOutcome(account.localId, success = false, message = message)
        }
    }

    /**
     * Best-effort persistence of a failure.
     *
     * These calls sit in catch blocks, outside any try. A throw from here would leave
     * [syncAccount] — documented as never throwing — and cancel the sibling accounts sharing
     * the [syncAll] scope, discarding their freshly fetched numbers.
     */
    private suspend fun recordFailure(accountId: String, message: String, requiresReauthentication: Boolean = false) {
        try {
            repository.saveFailure(accountId, message, nowMs(), requiresReauthentication)
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            // Nothing further to do: the error is already reflected in the returned outcome.
        }
    }

    /**
     * Returns usable credentials, refreshing first if the access token is at or near expiry.
     *
     * The refresh is guarded by a per-account mutex and the expiry is re-checked inside the
     * lock, so a caller that queued behind another refresh uses that result instead of
     * spending the (now-rotated) refresh token a second time.
     *
     * [rejectedAccessToken] is the access token a request actually presented and the provider
     * actually refused. It is stronger evidence than any clock, and it is the only thing that
     * rescues an account whose provider never states an expiry: `needsRefresh` is false when
     * `expiresAt` is null, so for such an account the proactive branch can never fire, and
     * before this every usage call 401'd for ever with no attempt to renew. A stored expiry
     * set too far in the future produced the same silence until that date.
     *
     * It is passed as the token itself rather than as a `force` flag on purpose. A concurrent
     * pass may already have saved a DIFFERENT token, which this request never presented and
     * nothing has refused — so it is offered first, and only the refused token costs a
     * rotation. Presenting an already-spent refresh token is how a provider revokes the whole
     * grant, so the difference is not academic. iOS decides it the same way.
     */
    /**
     * Runs [block] holding the per-reference credential lock.
     *
     * Sign-out has to take this too, and did not. `validCredentials` checks that the reference
     * still exists and then saves the rotated pair, both inside this lock — but a delete that
     * did not hold it could land in between, so the save put the credential the user had just
     * removed straight back. The store's own operations are each atomic; what needs
     * serialising is the sequence, and that means one lock shared by everything that changes a
     * credential's existence.
     *
     * Keyed by credential reference rather than by account, because two accounts can share
     * one reference and both would otherwise spend the same rotating refresh token.
     */
    suspend fun <T> withCredentialLock(reference: String, block: suspend () -> T): T {
        val mutex = mutexGuard.withLock { refreshMutexes.getOrPut(reference) { Mutex() } }
        return mutex.withLock { block() }
    }

    suspend fun validCredentials(
        account: ProviderAccount,
        rejectedAccessToken: String? = null,
    ): OAuthCredentials {
        val reference = account.credentialReference
        val current = credentialStore.load(reference)
            ?: throw ProviderException.Unauthorized("No stored credentials for this account")

        // A rejection overrides the clock, but only for the token that was rejected.
        val currentWasRejected = rejectedAccessToken != null &&
            current.accessToken == rejectedAccessToken
        if (!currentWasRejected && !current.needsRefresh(nowMs())) return current

        return withCredentialLock(reference) {
            val latest = credentialStore.load(reference)
                ?: throw ProviderException.Unauthorized("Account was removed while waiting to refresh")
            val latestWasRejected = rejectedAccessToken != null &&
                latest.accessToken == rejectedAccessToken
            // Whoever held the lock may have rotated already. If what they saved is not the
            // token this caller was refused, it is worth trying before spending another.
            if (!latestWasRejected && !latest.needsRefresh(nowMs())) return@withCredentialLock latest

            val provider = registry.forId(account.provider)
                ?: throw ProviderException.Unexpected("No provider for ${account.provider.id}")

            // The exchange and the write that follows it are one indivisible step.
            //
            // A rotating provider invalidates `latest` the moment it answers. If the worker is
            // cancelled while the response is in flight — a lost constraint, doze, the system
            // reclaiming it — the result can be discarded on the way out of the IO dispatcher,
            // or the coroutine can stop between receiving it and writing it. Either way the
            // provider has moved on and the store still holds a pair that is now dead: the
            // account is signed out, with nothing anywhere recording why, and the user's only
            // route back is to notice and reconnect it by hand.
            //
            // Cancellation is honoured everywhere else, and is still honoured here — just
            // AFTER the pair is safe. `withContext(NonCancellable)` covers only the exchange
            // and its persistence, so a cancelled sync stops at the next suspension point
            // rather than at the one place where stopping costs the account.
            withContext(NonCancellable) {
                val refreshed = provider.refresh(latest)
                // Written back only if the store still holds the very pair this exchange
                // started from — a compare-and-swap on content, not merely a presence check.
                //
                // Presence alone answered the wrong question. It catches the account being
                // REMOVED mid-flight, which is why it was there, and misses the account being
                // signed in AGAIN: the user reconnects the same provider while this refresh is
                // still in the air, the login writes fresh credentials under the same
                // reference, and this then overwrites them with a pair derived from the
                // session they just replaced. The symptom is the one users report as "I logged
                // in and it signed me straight back out", and it is the same failure as losing
                // a rotated pair, arriving from the other direction.
                //
                // `OAuthCredentials` is a data class, so identity here is the whole pair. If
                // anything at all changed underneath — a newer login, another refresh that
                // finished first — this result is stale by definition and is dropped rather
                // than allowed to win on arrival order.
                if (credentialStore.load(reference) == latest) {
                    credentialStore.save(reference, refreshed)
                }
                refreshed
            }
        }
    }
}

/** Message safe to show a user: never contains a token, URL, or raw provider body. */
fun ProviderException.userMessage(): String = when (this) {
    // The evaluator's constant, not a second copy of the sentence. See
    // `NotificationEvaluator.SIGN_IN_EXPIRED_MESSAGE` for the notification a divergence here
    // silently disabled on the other platform.
    is ProviderException.Unauthorized -> NotificationEvaluator.SIGN_IN_EXPIRED_MESSAGE
    is ProviderException.Forbidden -> "Access denied for this account"
    is ProviderException.RateLimited -> "Rate limited — try again shortly"
    is ProviderException.ServerError -> "Provider is having trouble ($statusCode)"
    // The client names the host and the kind of failure; the generic line is only the fallback.
    is ProviderException.Offline -> message ?: "No network connection"
    is ProviderException.MalformedPayload -> "Unexpected response from provider"
    is ProviderException.LoginCancelled -> "Sign-in was cancelled"
    is ProviderException.Unexpected -> "Refresh failed"
}

File: ios/UsageLimitsKit/Sources/UsageLimitsKit/Sync/SyncEngine.swift
import Foundation

/// What a provider can do for one account.
public protocol SyncProvider: Sendable {

    /// The id this provider is registered under. Accounts name the provider they want through an
    /// registered provider, so a provider the engine was never
    /// given fails that one account, never the whole batch.
    var providerID: String { get }

    /// Reads the account's current usage, using the credentials the provider itself issued.
    func fetchUsage(credentials: OAuthCredentials, attributes: [String: String]) async throws -> UsageResult

    /// Exchanges a refresh token for a new set. Throws if it cannot.
    ///
    /// Most of these providers retire the refresh token on every exchange, which makes an exchange
    /// a one-shot operation: the loser of two concurrent exchanges is left holding a token the
    /// provider has already invalidated, and the account then reads as revoked when nothing is
    /// wrong with it.
    func refresh(credentials: OAuthCredentials) async throws -> OAuthCredentials
}

/// What one account's refresh came to.
public enum SyncOutcome: Sendable {

    /// The account's usage was fetched.
    case success(accountID: String, result: UsageResult)

    /// The account could not be refreshed. The message is built from the error and from
    /// identifiers the engine chose; credential material is never interpolated into it, so an
    /// outcome is safe to display exactly as it stands.
    case failure(accountID: String, message: String)
}

/// Receives every outcome as it happens, successes and failures alike, so accounts settle on the
/// caller's screen the moment they finish.
public protocol SyncSink: Sendable {

    /// Deliberately non-throwing: a fault while persisting an outcome must never be turned back
    /// into an account failure, so a sink with problems owns them itself.
    func record(_ outcome: SyncOutcome, at time: Date) async
}

/// Refreshes quota for several AI-subscription accounts.
///
/// The engine is an actor for the sake of a single piece of state: the map of in-flight refreshes
/// keyed by credential reference. Everything an individual account does could run anywhere; the
/// guarantee that only one exchange per reference is ever in flight needs shared, serialised
/// state, and the actor is where it lives.
public actor SyncEngine {

    /// The `ProviderAccount.attributes` key under which an account names the provider that serves
    /// it. It lives in the attributes because the account type carries no provider field of its
    /// own.

    /// A pair this close to its recorded expiry is refreshed now rather than during the fetch that
    /// is about to run: a token that expires while the request is in flight fails the fetch for no
    /// real reason. Both expiry checks use the same leeway, so a pair judged worth refreshing
    /// outside the serial section is still judged worth refreshing inside it.
    public static let expiryLeeway: TimeInterval = 60

    private let providers: [String: any SyncProvider]
    private let credentials: any CredentialStore
    private let sink: any SyncSink
    private let now: @Sendable () -> Date

    /// One exchange in progress per credential reference. A caller that finds an entry awaits it
    /// rather than starting a rival exchange for the same rotating refresh token.
    private var refreshes: [String: Task<Exchange, any Error>] = [:]

    /// What one trip through `renewedCredentials` produced.
    ///
    /// `exchanged` distinguishes "I spent a rotation and here is the new pair" from "somebody
    /// else had already refreshed this reference, so here is what they saved". A caller that
    /// forced the exchange because the provider REJECTED the token needs to tell those apart:
    /// joining a non-forced exchange that decided nothing needed doing hands back the same dead
    /// token, and the account is then reported as revoked when it is merely stale.
    private struct Exchange: Sendable {
        let credentials: OAuthCredentials
        let exchanged: Bool
    }

    /// How many times a rotated pair is offered to the store before giving up on writing it.
    ///
    /// Retried rather than surrendered, because by the time the write is attempted the OLD
    /// refresh token is already dead: a failed save does not leave things as they were, it loses
    /// the account. The keychain items are `AfterFirstUnlockThisDeviceOnly`, so the usual
    /// transient cause — a locked device — cannot arise; what remains is worth a second attempt
    /// and not worth a long one.
    private static let saveAttempts = 3

    public init(providers: [String: any SyncProvider],
                credentials: any CredentialStore,
                sink: any SyncSink,
                now: @Sendable @escaping () -> Date = { Date() }) {
        self.providers = providers
        self.credentials = credentials
        self.sink = sink
        self.now = now
    }

    /// Refreshes every account given and returns the outcomes in the same order.
    ///
    /// The accounts run together, so a slow one delays only itself, and each is isolated: anything
    /// a single account throws becomes that account's failure and nothing else's, and every other
    /// account still syncs and is recorded — one bad account never aborts the batch. The one error
    /// that escapes is `CancellationError`: a cancelled run is not a broken account, so it
    /// propagates to the caller instead of being recorded as a failure.
    public func sync(accounts: [ProviderAccount]) async throws -> [SyncOutcome] {
        try Task.checkCancellation()

        var outcomes = Array<SyncOutcome?>(repeating: nil, count: accounts.count)

        try await withThrowingTaskGroup(of: (Int, SyncOutcome).self) { group in
            for (index, account) in accounts.enumerated() {
                group.addTask {
                    let outcome = try await self.run(account)
                    return (index, outcome)
                }
            }

            for try await (index, outcome) in group {
                outcomes[index] = outcome
            }
        }

        // A child that throws stops the group from returning at all, so by the time control
        // reaches here every slot has been filled exactly once.
        return outcomes.map { $0! }
    }

    /// Carries one account from outcome to recorded outcome, and returns it so the batch can be
    /// answered in the order it was asked for. Delivery is per account rather than at the end of
    /// the batch, so an outcome reaches the sink the moment it settles, whatever its neighbours
    /// are doing.
    private func run(_ account: ProviderAccount) async throws -> SyncOutcome {
        let outcome = try await isolatedOutcome(for: account)
        try await record(outcome)
        return outcome
    }

    /// The failure boundary for a single account.
    ///
    /// Every error the pipeline raises becomes this account's failure alone; the neighbouring
    /// accounts carry on and are recorded regardless. `CancellationError` is the single exception:
    /// it stands for the caller giving up, not for an account going wrong, and recording it as a
    /// failure would paint a healthy account red because a screen was closed.
    private func isolatedOutcome(for account: ProviderAccount) async throws -> SyncOutcome {
        do {
            let result = try await usage(for: account)
            return .success(accountID: account.id, result: result)
        } catch let cancelled as CancellationError {
            throw cancelled
        } catch {
            return .failure(accountID: account.id, message: Self.message(for: error))
        }
    }

    /// One account's quota refresh: resolve the provider, arrive at a credential pair that is not
    /// known to be expired, fetch the usage.
    private func usage(for account: ProviderAccount) async throws -> UsageResult {
        try Task.checkCancellation()

        // The account's own typed provider, not a stringly-typed attribute alongside it. Reading
        // an attribute meant a perfectly well-formed account — one whose `provider` says exactly
        // which provider it is — failed with "does not name a provider" because a second, parallel
        // copy of that fact had not been filled in. Two sources of truth for one thing, where
        // forgetting the redundant one costs the user a silently broken account.
        let providerID = account.provider.rawValue
        guard let provider = providers[providerID] else {
            // An id the engine has no provider for is stale configuration, not a reason to stop the
            // process: this account fails and the rest of the batch carries on.
            throw SyncEngineError.providerNotRegistered(providerID)
        }

        guard let stored = try await credentials.load(reference: account.credentialReference) else {
            // Nothing can be exchanged without a pair to exchange, so the account fails here
            // rather than somewhere deeper inside the provider.
            throw SyncEngineError.credentialsAbsent
        }

        let usable: OAuthCredentials
        if stored.isExpired(now: now(), leeway: Self.expiryLeeway) {
            usable = try await renewedCredentials(
                reference: account.credentialReference,
                provider: provider
            ).credentials
        } else {
            // "Not known to be expired" deliberately includes an absent expiry — refreshing such an
            // account on every call would burn a rotating refresh token each time.
            usable = stored
        }

        // If the run was cancelled while the credentials were being sorted out, give up before the
        // network rather than after it.
        try Task.checkCancellation()

        do {
            return try await provider.fetchUsage(credentials: usable, attributes: account.attributes)
        } catch ProviderError.unauthorised {
            // The reactive half of the refresh policy, and the half that makes the proactive one
            // safe to keep conservative.
            //
            // "Not known to be expired" includes an absent expiry, because refreshing on every
            // call would burn a rotating refresh token each time. On its own that leaves an
            // account whose provider never states an expiry to 401 for ever, silently — the
            // proactive branch can never fire for it. So a rejection is what triggers the
            // exchange instead: the provider itself has said the token is dead, which is better
            // evidence than any clock.
            //
            // What is passed is the access token this request actually tried, not a bare
            // "force". The difference matters twice. A concurrent pass may already have saved a
            // DIFFERENT token, which this request has not tried and which no evidence says is
            // dead — so it is worth one attempt before spending a rotation. And the retry
            // condition below used to be `!renewed.exchanged`, which spent a second rotation
            // even when the pair handed back was a different, freshly saved one. Identity of the
            // rejected token answers both questions exactly; `exchanged` only approximated them.
            //
            // Once only. If the pair that comes back is rejected too, the credential is
            // genuinely revoked and retrying is just a second way to fail.
            let renewed = try await renewedCredentials(
                reference: account.credentialReference,
                provider: provider,
                rejectedAccessToken: usable.accessToken
            )
            try Task.checkCancellation()
            return try await provider.fetchUsage(
                credentials: renewed.credentials, attributes: account.attributes)
        }
    }

    /// Returns a pair stored under `reference` that is not known to be expired, with at most one
    /// exchange for the reference in flight at a time.
    ///
    /// The lock is keyed by reference, not by account, because two accounts can share one
    /// credential reference and both would otherwise spend the same rotating refresh token. A
    /// caller that finds an exchange already running awaits it.
    ///
    /// The expiry is re-decided inside the serial section, against a pair re-read from the store.
    /// That re-read is what stops a caller arriving just after an earlier exchange finished from
    /// exchanging again immediately: the earlier exchange has already saved a fresh pair, and the
    /// re-read finds it. Trusting the copy the caller loaded before it waited is precisely the
    /// double refresh that used to strand accounts.
    /// - Parameter force: exchange even when the stored pair does not look expired. Set only by
    ///   the rejection path, where the provider has already said the token is dead.
    private func renewedCredentials(reference: String,
                                    provider: any SyncProvider,
                                    rejectedAccessToken: String? = nil) async throws -> Exchange {
        // A loop rather than a single join. A caller whose token was REJECTED can join an
        // exchange that was started proactively, decided nothing needed doing, and handed back
        // the very token the provider just refused. Retrying the endpoint with it would report a
        // healthy account as revoked. By the time that value is available the joined task's
        // `defer` has already cleared its entry, so coordinating again either starts a real
        // exchange or joins a later one — it cannot rejoin the same no-op.
        while let running = refreshes[reference] {
            let result = try await running.value
            if let rejectedAccessToken,
               !result.exchanged,
               result.credentials.accessToken == rejectedAccessToken {
                continue
            }
            return result
        }

        // The exchange runs as a deliberately unstructured task, which the engine never cancels:
        // an exchange abandoned between the rotation and the save leaves the old refresh token
        // dead and the new one unwritten, losing the account outright. For the same reason the
        // body performs no cancellation check — once begun, an exchange runs to its save.
        let task = Task<Exchange, any Error> {
            // The entry lasts exactly as long as the exchange: a late arrival either joins this
            // exchange or re-reads a store this exchange has already written to.
            defer { self.refreshes[reference] = nil }

            guard let current = try await self.credentials.load(reference: reference) else {
                throw SyncEngineError.credentialsAbsent
            }

            if let rejectedAccessToken {
                if current.accessToken != rejectedAccessToken {
                    // The store has moved on: a concurrent pass saved a different token, and
                    // nothing has rejected THAT one. Offer it before spending a rotation.
                    return Exchange(credentials: current, exchanged: false)
                }
                // Otherwise the stored token is the rejected one, and its expiry is irrelevant —
                // the provider has already said it is dead, which beats any clock.
            } else if !current.isExpired(now: self.now(), leeway: Self.expiryLeeway) {
                // An exchange for this reference has already finished; the saved pair is fresh,
                // and spending another rotation here is the very bug this routine exists to
                // prevent.
                return Exchange(credentials: current, exchanged: false)
            }

            let response = try await provider.refresh(credentials: current)

            // A token response is not a complete credential. Providers that do not rotate their
            // refresh token omit it from the response entirely, and storing the response as-is
            // would delete the refresh token this account depends on — losing an account that
            // nothing had revoked, which looks exactly like a revocation.
            let renewed = current.merging(refreshed: response)

            // Persisted before it is used — by this exchange or by any account waiting on it. If
            // the process dies between the exchange and the fetch, the old refresh token is
            // already dead, and a pair that was never written back loses the account entirely.
            //
            // Which is also why a failing save is retried and then, if it still will not write,
            // NOT allowed to fail the refresh. The rotation has already happened: the token in
            // the store is dead whatever this code does next. Throwing here would discard the
            // only working pair that exists and turn a storage fault into a signed-out paid
            // account within the same second. Returning it means this session keeps working,
            // and the sign-out is deferred to the next launch rather than caused now.
            await self.persist(renewed, reference: reference)
            return Exchange(credentials: renewed, exchanged: true)
        }

        // Nothing between the look-up above and this store suspends, so on this actor the
        // check-and-register is atomic: at most one exchange is ever registered per reference.
        refreshes[reference] = task
        return try await task.value
    }

    /// Writes a rotated pair, retrying, and never throwing. See the call site for why a failure
    /// here must not fail the refresh.
    private func persist(_ credentials: OAuthCredentials, reference: String) async {
        for attempt in 1...Self.saveAttempts {
            do {
                // Update-only, never insert. A rotation that started before the user removed
                // the account finishes afterwards, and `save` would put the deleted credential
                // straight back: the account is gone from the list while a usable token stays
                // in the keychain with nothing left to clean it up. A false return means
                // removal won, which is a correct outcome and not a failure to retry.
                _ = try await self.credentials.updateIfPresent(credentials, reference: reference)
                return
            } catch {
                guard attempt < Self.saveAttempts else { return }
                // A short, fixed pause. The plausible causes are momentary; a long backoff would
                // hold the exchange lock open while every other account waits on it.
                try? await Task.sleep(nanoseconds: 100 * 1_000_000)
            }
        }
    }

    /// Delivers one outcome to the sink, best-effort.
    ///
    /// The sink cannot throw, so persisting an outcome can never fail the account it belongs to —
    /// that is what makes the delivery best-effort. The outcome is delivered even if the run has
    /// been cancelled in the meantime, because the work is done and the sink should not lose it;
    /// cancellation that lands while the delivery is in flight must still escape, so it is raised
    /// once the sink has had its chance.
    private func record(_ outcome: SyncOutcome) async throws {
        await sink.record(outcome, at: now())
        try Task.checkCancellation()
    }

    /// Renders a caught error as an outcome message.
    ///
    /// The engine builds messages from the error itself and from identifiers it chose; credential
    /// material is never interpolated, so an outcome is safe to show the user as it stands.
    private static func message(for error: any Error) -> String {
        if let described = (error as? LocalizedError)?.errorDescription, !described.isEmpty {
            return described
        }
        return String(describing: error)
    }

    /// Faults the engine raises for a single account. The payloads are identifiers, never
    /// credential material, so every message built from them is safe to display.
    private enum SyncEngineError: LocalizedError, Sendable {

        /// `nil` when the account names no provider at all.
        case providerNotRegistered(String?)
        case credentialsAbsent

        var errorDescription: String? {
            switch self {
            case .providerNotRegistered(nil):
                return "The account does not name a provider."
            case .providerNotRegistered(let providerID?):
                return "No provider is registered for provider id \"\(providerID)\"."
            case .credentialsAbsent:
                return "No credentials are stored for the account."
            }
        }
    }
}